First pass at 'paasta maintenance' command

nhandler opened the pull request:
This is a first pass at a `paasta maintenace` command. I need to think a bit more about how to properly do itests in a useful way, but there are some unit tests.

Connect to #376

solarkennedy commented:
We don't usually put this kind of script as a subcommand of the "paasta" command. Mostly because the paasta super-command is very user-centric, and this not something a user would ever run. It makes assumptions that it can make API calls, has auth? and can directly connect to the mesos api, which none of the other paasta commands do.

Please start his off as like, paasta_maintenance? Like was your expectation would be that you could run this from anywhere, or from the local slave? Or from the master?

nhandler commented:
Ok, I have gone ahead and moved this to a `paasta_maintenance` script.

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,244 @@
+    try:
+        leader = get_mesos_leader()
+    except MesosMasterConnectionError:
+        print "Run this script from a mesos master!"
```
This is surprising - I expected the workflow would be 'log onto a slave, mark as in maintenance' rather than 'discover where the master is, ssh in, mark a slave as down'. I'm not saying this isn't the right way, just not the way I immediately expected.

nhandler replied:
Part of the reason for this is that you can mark a host as being up/down without the box itself being available. i.e. when imaging some new PaaSTA slaves, we've been marking them as down so that they do not immediately register when they come up (we want to wait until Puppet is done running). If we had to SSH to the slave to do this, it would not be possible.

Rob-Johnson replied:
ok, then how about a  ``--hostname`` option that defaults to the current host, rather than constraining this to *only* run on a master. Another use case for this might be 'log into a box, mark as down, drain, restart docker, mark as up', which becomes more of a pain if you have to perform some commands on the master and some on the slave. Is there any technical constraint to *only* allow it on the master, rather than either?

Rob-Johnson replied:
bump

nhandler replied:
That could work. I'll need to flip from mesos_tools.py:get_mesos_leader()  to using mesos.cli, but that shouldn't be too bad.

Rob-Johnson commented:
There are a couple of things that stand out here:

- If this is executed on the master, how will you drain the box? Will there be another process on each slave that checks whether it is in drain mode, and acts accordingly?

- There doesn't seem to be any error handling - requests will throw descriptive exceptions, but there is no logic to handle them here.

- There is an awful lot of repetition here. I'd encourage you try and refactor the api client to avoid this. 

Here's a start on something you could use - note that the final request execution logic is handled in one place, rather than *every* function, and a "client" of sorts is built for every HTTP resource. The final endpoint (like ``/status``)  is then just executing the client for a given resource. (Note I've not tested any of this code, so it's probably buggy - this is just an idea of what It might look like)

```
def base_api():
    leader = get_mesos_leader()
    def execute_request(method, endpoint, **kwargs):
        endpoint = "http://leader:PORT/%s" % endpoint
        s = Session()
        req = Request(method, endpoint, **kwargs)
        prepped = s.prepare_request(req)
        try:
            resp = s.send(prepped,
               stream=stream,
               verify=verify,
               proxies=proxies,
               cert=cert,
               timeout=timeout
            )
            resp.raise_for_status()
         except Exception as e:
            print("Error executing api request calling %s . Got response code %d with error %s" %(endpoint, resp.status_code, e)
    return execute_request


def master_api_client():
    def execute_master_api_request(method, endpoint)
        base_api_client = base_api()
        return base_api_client(method, "master/%s" % endpoint, **kwargs)
    return execute_master_api_request


def maintainance_api_client():
    def execute_schedule_api_request(method, endpoint)
        master_api_client = maintenance_api_client()
        return base_api_client(method, "schedule/%s" % endpoint, **kwargs)
    return execute_schedule_api_request


def get_schedule_client():
    def execute_schedule_api_request(method, endpoint)
        master_api_client = maintenance_api_client()
        return master_api_client(method, "schedule/%s" % endpoint, **kwargs)
    return execute_schedule_api_request


def get_maintenance_schedule():
    client_fn = get_schedule_client()
    return client_fn("GET", "")


def get_maintenance_status():
   client_fn = get_schedule_client()
   return client_fn("GET", "status")
```

nhandler commented:
@Rob-Johnson I've attempted to refactor the code a bit based on your feedback. It could still probably use more error handling (I'm working on adding it), but I wanted to get this change up for more review.

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,276 @@
+def get_machine_ids(hostnames):
+    machine_ids = []
+    for hostname in hostnames:
+        machine_id = dict()
```
make this a named tuple rather than a dictionary.

```
from collections import namedtuple
MachineId = namedtuple('MachineId', 'hostname, ip')

machine_id = MachineId(hostname='foo', ip='bar')
x = machine_id.hostname
y = machine_id.bar
```

nhandler replied:
The dictionary was used because this gets converted to JSON further down.

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,276 @@
+    machine_ids = []
+    for hostname in hostnames:
+        machine_id = dict()
+        # This is to allow specifying a hostname as "hostname|ipaddr"a
```
why do we need this? are you that concerned about the overhead of a dns request?

nhandler replied:
I'm not worried at all about the overhead. We ran into a problem a while ago where we updated DNS and I marked a box as being 'down' in preparation for it being imaged. However, there was some error with the DNS change that required us to change where the hostname pointed. Once the revised DNS change was pushed, I had no easy way to mark the old host/ip mapping as being up (since the maintenance script was resolving the hostname to a different IP).

The other case where this is useful is if you want to mark a box as 'down' prior to it even existing in DNS.

Rob-Johnson replied:
man, that sounds like a bit of a corner case you're optimising for. If you insist on it, at least add some docs to the help text to tell people about it.

nhandler replied:
I really don't think this situation will be that uncommon. I am totally up for other methods of allowing the user to specify hostname/ip pairs rather than the style I have now, but I think having this functionality is necessary. I will add a note to the argparse help for this argument.

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,276 @@
+    return client_fn(method="GET", endpoint="status")
+
+
+def timedelta_type(value):
```
this isn't really a type, is it?

nhandler replied:
We used something similar in https://github.com/Yelp/paasta/blob/b89717a110d1bbf0025f7cb6e254507e7846eb4a/paasta_tools/contrib/delete_old_marathon_deployments.py#L42 . It isn't an official type. But per https://docs.python.org/3/library/argparse.html#type , we can use this function with argparse to handle type-checking and type conversion at the argparse level.

Rob-Johnson replied:
ah right, ok. heh, TIL!

ok, in that case... can you be more explicit? call it ``parse_timedelta`` or something. Also, you should probably catch the exception that'll be thrown by ``timeparse.timeparse(value)`` when this isn't legit and throw the correct ``argparse.ArgumentTypeError`` so that the output is formatted correctly by argparse

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,276 @@
+        status()
+        schedule()
+    else:
+        print "Unknown Action: %s" % action
```
raise a ValueError here

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,276 @@
+    return machine_ids
+
+
+def build_maintenance_schedule_payload(hostnames, start, duration, drain=True):
```
docstrings please

Rob-Johnson commented:
please add docstrings to your functions for readability, too

nhandler commented:
I think I've addressed all of the feedback. Our general_itests currently fail due to importing `paasta_tools.mesos_tools` but never using it (we need the side-effect of it setting an environment variable for `mesos.cli`). I plan to fix #498, which should resolve this issue.

nhandler commented:
#498 has been merged. I've updated this PR to use `paasta_tools.mesos_tools.get_mesos_leader` instead of using `mesos.cli` directly.

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,367 @@
+    maintenance API requests.
+    :param mesos_secrets: optional argument specifying the path to the file containing the mesos-slave credentials
+    """
+    with open(mesos_secrets) as data_file:
```
need some handling a missing file here?

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,367 @@
+    """
+    with open(mesos_secrets) as data_file:
+        data = json.load(data_file)
+    username = data['principal']
```
key errors?

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,367 @@
+    """
+    payload = build_maintenance_schedule_payload(hostnames, start, duration, drain=True)
+    client_fn = get_schedule_client()
+    print client_fn(method="POST", endpoint="", data=json.dumps(payload)).text
```
given that you're not raising the exception in base_api (line 79), I think you might get caught out here if there is an exception in the request, the client_fn won't return anything.

nhandler replied:
I guess I'll update `base_api()` to re-raise the `HTTPError` with a friendlier message.

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,367 @@
+        status()
+        schedule()
+    else:
+        raise ValueError("Unknown Action: %s" % action)
```
will you actually get here? you'll return on line 342 won't you?

nhandler replied:
Correct. I added the check at 342 to fail a bit earlier. after this else statement was already here. I'll remove it.

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,367 @@
+        'You can specify <hostname>|<ip> to avoid querying DNS to determine the corresponding IP.',
+    )
+    options = parser.parse_args()
+    return options
```
nitpick, but why not just ``return parser.parse_args()`` ?

nhandler replied:
I think I borrowed this from another script. I'll change it.

Rob-Johnson commented on paasta_tools/paasta_maintenance.py:
```diff
@@ -0,0 +1,367 @@
+        req = Request(method, url, **kwargs)
+        prepared = s.prepare_request(req)
+        try:
+            resp = s.send(prepared,
```
indentation looks a bit odd here

nhandler replied:
I believe this is actually valid and works. I'll move `prepared` to its own line, which will allow me to make the overall indentation look a bit more as you were probably expecting.

Rob-Johnson commented:
you're still missing a lot of ``:returns:`` in your docstrings

Rob-Johnson commented:
fix and ship

Rob-Johnson commented:
oh man, what's up with the giant diff? have you rebased or something?
